Sources/Datadog/URLSessionAutoInstrumentation/Interception/AppStateListener.swift

```diff
@@ -0,0 +1,167 @@
+        taken.initialState = Snapshot(
+            isInBackground: isInBackground(at: range.lowerBound),
+            date: range.lowerBound
+        )
+        // move final state to upperBound
+        taken.finalState = Snapshot(
+            isInBackground: isInBackground(at: range.upperBound),
+            date: range.upperBound
+        )
+        // filter changes outside of the range
+        taken.changes = taken.changes.filter { range.contains($0.date) }
+        return taken
+    }
+
+    var foregroundDuration: TimeInterval {
+        var duration: TimeInterval = 0.0
+        var lastForegroundStartDate: Date?
+        let allEvents = [initialState] + changes.sorted() + [finalState]
+        for event in allEvents {
+            if let startDate = lastForegroundStartDate {
+                duration += event.date.timeIntervalSince(startDate)
+            }
+            if event.isInBackground {
+                lastForegroundStartDate = nil
+            } else {
+                lastForegroundStartDate = event.date
+            }
+        }
+        return duration
+    }
+
+    var didRunInBackground: Bool {
+        return initialState.isInBackground || finalState.isInBackground
+    }
+
+    private func isInBackground(at date: Date) -> Bool {
+        if date <= initialState.date {
+            // we assume there was no change before initial state
+            return initialState.isInBackground
+        } else if finalState.date <= date {
+            // and no change after final state
+            return finalState.isInBackground
+        }
+        var isInBackground = initialState.isInBackground
+        for change in changes {
+            if date < change.date {
+                break
+            }
+            isInBackground = change.isInBackground
+        }
+        return isInBackground
+    }
+}
+
+internal protocol AppStateListening {
+    var appStateHistory: AppStateHistory { get }
+}
+
+extension AppStateListening {
+    static func build() -> AppStateListening {
+        #if canImport(UIKit)
+        return AppStateListener()
+        #else
+        // keep #else block at minimum since it's hard to test
+        return NonUIKitAppStateListener()
+        #endif
+    }
+}
+
+#if canImport(UIKit)
+import class UIKit.UIApplication
+
+internal class AppStateListener: AppStateListening {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    private static var isBackground: Bool {
+        return UIApplication.managedShared?.applicationState == .background
+    }
+
+    private static func currentState(with date: Date) -> Snapshot {
+        return Snapshot(isInBackground: AppStateListener.isBackground, date: date)
+    }
+
+    private let dateProvider: DateProvider
+    private var _appStateHistory: AppStateHistory
+    var appStateHistory: AppStateHistory {
+        _appStateHistory.finalState = AppStateListener.currentState(with: dateProvider.currentDate())
+        return _appStateHistory
+    }
```

Comment:
good point 👍  
`appStateHistory` shouldn't be called from non-main thread as it calls `UIApplication.shared`, otherwise `Main Thread Checker` gives warnings. i need to ensure that.

i thought `UIApplication.back/foregroundNotifications` are always posted in main thread but i can't find any mention in docs. i need to ensure main thread there too.